code/game/objects/machinery/door_display/door_display.dm

```diff
@@ -13,99 +13,102 @@
 		if(SOUTH)
-			pixel_y = 32
+			pixel_y += 32
 		if(EAST)
-			pixel_x = 32
+			pixel_x += 32
 		if(WEST)
-			pixel_x = -32
+			pixel_x += -32
 
 	return INITIALIZE_HINT_LATELOAD
 
 /obj/machinery/door_display/LateInitialize()
 	for(var/obj/machinery/door/D in GLOB.machines)
-		if (D.id == id)
+		if(D.id == id)
 			targets += D
+	if(has_flash)
+		for(var/obj/machinery/flasher/flasher in GLOB.machines)
+			if(flasher.id == id)
+				targets += flasher
 
 	if(!length(targets))
 		machine_stat |= BROKEN
 	update_icon()
 
-// open/closedoor checks if door_display has power, if so it checks if the
-// linked door is open/closed (by density) then opens it/closes it.
-
-// Opens and locks doors, power check
-/obj/machinery/door_display/proc/open_door()
-	if(machine_stat & (NOPOWER|BROKEN))
-		return FALSE
-
-	for(var/obj/machinery/door/D in targets)
-		if(!D.density)
-			continue
-		D.open()
-
-	open = TRUE
-
-	return TRUE
-
-
-// Closes and unlocks doors, power check
-/obj/machinery/door_display/proc/close_door()
-	if(machine_stat & (NOPOWER|BROKEN))
-		return FALSE
-
-	for(var/obj/machinery/door/D in targets)
-		if(D.density)
+/obj/machinery/door_display/proc/toggle_doors(state)
+	for(var/obj/machinery/door/door in targets)
+		if(state) //true for open
+			door.open()
+		else
+			door.close()
+	open = state
+
+/obj/machinery/door_display/proc/toggle_shutters(state)
+	if(!state)
+		toggle_doors(FALSE)
+	for(var/obj/machinery/door/poddoor/door in targets)
+		if(door.density == !state)
 			continue
-		D.close()
-
-	open = FALSE
-
-	return TRUE
-
-
-/obj/machinery/door_display/interact(mob/user)
-	. = ..()
-	if(.)
-		return
-
-	var/data
-	data += "<HR>Linked Door:</hr>"
-	data += " <b> [id]</b><br/>"
-
-	if(open)
-		data += "<a href='?src=\ref[src];open=0'>Close Door</a><br/>"
-	else
-		data += "<a href='?src=\ref[src];open=1'>Open Door</a><br/>"
-
-	var/datum/browser/popup = new(user, "computer", "<div align='center'>Door controls</div>", 300, 240)
-	popup.set_content(data)
-	popup.open()
+		if(state) //true for open
+			door.open()
+		else
+			door.close()
+	open_shutter = state
 
+/obj/machinery/door_display/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if (!ui)
+		ui = new(user, src, "DoorDisplay", "[name]")
+		ui.open()
 
-/obj/machinery/door_display/Topic(href, href_list)
+/obj/machinery/door_display/ui_act(action, list/params, datum/tgui/ui)
 	. = ..()
 	if(.)
 		return
-
-	if(href_list["open"])
-		if (open)
-			close_door()
-		else
-			open_door()
-		update_icon()
-
-	updateUsrDialog()
-
+	if(machine_stat & (NOPOWER|BROKEN))
+		return FALSE
+	switch(action)
+		if("door")
+			var/state = text2num(params["state"])
+			if(!isnum(state))
+				return FALSE
+			toggle_doors(state)
+			return TRUE
+		if("shutter")
+			var/state = text2num(params["state"])
+			if(!isnum(state))
+				return FALSE
+			toggle_shutters(state)
+			return TRUE
+		if("flash")
+			for(var/obj/machinery/flasher/flasher in targets)
```

Comment:
why is this typechecking

Reply:
targets is a list of all machines linked to it for some reason, remnant of the old code